arangod/RocksDBEngine/RocksDBTransactionState.cpp

```diff
@@ -491,226 +282,67 @@ Result RocksDBTransactionState::abortTransaction(transaction::Methods* activeTrx
-    return engine.db()->GetLatestSequenceNumber();
-  }
-  TRI_ASSERT(false);
-  THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "No snapshot set");
-}
-
-/// @brief acquire a database snapshot
-bool RocksDBTransactionState::setSnapshotOnReadOnly() {
-  if (_readSnapshot == nullptr && isReadOnlyTransaction()) {
-    TRI_ASSERT(_rocksTransaction == nullptr);
-    TRI_ASSERT(isSingleOperation());
-    auto& selector = vocbase().server().getFeature<EngineSelectorFeature>();
-    auto& engine = selector.engine<RocksDBEngine>();
-    _readSnapshot = engine.db()->GetSnapshot();
-    return true;
-  }
-  return false;
-}
-
-Result RocksDBTransactionState::triggerIntermediateCommit(bool& hasPerformedIntermediateCommit) {
-  TRI_ASSERT(!hasPerformedIntermediateCommit);
-
-  TRI_IF_FAILURE("FailBeforeIntermediateCommit") {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-  }
-  TRI_IF_FAILURE("SegfaultBeforeIntermediateCommit") {
-    TRI_TerminateDebugging("SegfaultBeforeIntermediateCommit");
-  }
+  Result result = _rocksMethods->addOperation(cid, revisionId, operationType, hasPerformedIntermediateCommit);
+
+  if (result.ok()) {
+    auto tcoll = static_cast<RocksDBTransactionCollection*>(findCollection(cid));
+    if (tcoll == nullptr) {
+      std::string message = "collection '" + std::to_string(cid.id()) +
+                            "' not found in transaction state";
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, message);
+    }
 
-  TRI_ASSERT(!hasHint(transaction::Hints::Hint::SINGLE_OPERATION));
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-  LOG_TOPIC("0fe63", DEBUG, Logger::ENGINES) << "INTERMEDIATE COMMIT!";
-#endif
+    // should not fail or fail with exception
+    tcoll->addOperation(operationType, revisionId);
 
-  Result res = internalCommit();
-  if (res.fail()) {
-    // FIXME: do we abort the transaction ?
-    return res;
-  }
+    // clear the query cache for this collection
+    auto queryCache = arangodb::aql::QueryCache::instance();
 
-  hasPerformedIntermediateCommit = true;
-  ++statistics()._intermediateCommits;
-  
-  // reset counters for DML operations, but intentionally don't reset
-  // the commit counter, as we need to track if we had intermediate commits
-  _numInserts = 0;
-  _numUpdates = 0;
-  _numRemoves = 0;
-  _numLogdata = 0;
-
-  TRI_IF_FAILURE("FailAfterIntermediateCommit") {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
-  }
-  TRI_IF_FAILURE("SegfaultAfterIntermediateCommit") {
-    TRI_TerminateDebugging("SegfaultAfterIntermediateCommit");
+    if (queryCache->mayBeActive() && tcoll->collection()) {
+      queryCache->invalidate(&_vocbase, tcoll->collection()->guid());
+    }
   }
-
-  createTransaction();
-  _rocksReadOptions.snapshot = _rocksTransaction->GetSnapshot();
-  TRI_ASSERT(_readSnapshot != nullptr);  // snapshots for iterators
-  TRI_ASSERT(_rocksReadOptions.snapshot != nullptr);
-  return TRI_ERROR_NO_ERROR;
+  return {};
```

Comment:
The value of `result` should be returned here. Otherwise we may have a failure in `result` but not react on it.
```suggestion
  return result;
```